src/internal/executor/iexecutor.cpp

```diff
@@ -27,35 +27,34 @@
 
 namespace srf::internal::executor {
 
-IExecutor::IExecutor(std::shared_ptr<Options> options) : m_impl(make_executor(std::move(options))) {}
-IExecutor::IExecutor(std::unique_ptr<system::ISystem> system) : m_impl(make_executor(system::System::unwrap(*system)))
+IExecutor::IExecutor(std::unique_ptr<Executor> impl) : m_impl(std::move(impl))
 {
     CHECK(m_impl);
 }
+IExecutor::IExecutor(std::shared_ptr<Options> options) : IExecutor(make_executor(std::move(options))) {}
+IExecutor::IExecutor(std::unique_ptr<system::ISystem> system) :
+  IExecutor(make_executor(system::System::unwrap(*system)))
+{}
 
 IExecutor::~IExecutor() = default;
 
 void IExecutor::register_pipeline(std::unique_ptr<internal::pipeline::IPipeline> pipeline)
 {
-    CHECK(m_impl);
     m_impl->register_pipeline(std::move(pipeline));
 }
 
 void IExecutor::start()
 {
-    CHECK(m_impl);
```

Comment:
I think these need to stay.

```
srf::Executor executor(options)
parent->set_executor(std::move(executor));
executor.start(); // segfaults without checks; checks provide clearer reason for failure
```

Reply:
This seems more like a code smell than desired behavior. When and why would we want to swap out the executor, and why wouldn't set_executor be responsible for the `CHECK`?

Reply:
This is just move semantics. 

I don't think we can mark Executor as not-moveable.

Reply:
It appears we can, and probably should? I don't see why something as high level as Executor would need to be moved when we're heap allocating it and keeping it in a `unique_ptr`.

Reply:
it's a `unique_ptr`, It's privately owned, and we're not moving it. Why are we checking that it's not null every time we use it? We know it's not null because we're not nullifying it, we're not moving it, and no-one else has access to it. I'm confused why this check is necessary.

Reply:
it's owned by a class that is moveable both by construction and assignment.

```
auto exec2 = std::move(exec1)
```

`exec2` now owns the unique_ptr held by `exec1`; `exec1`'s unique_ptr is null.

Reply:
We discussed offline, and it sounds like we're good to go with these changes, assuming Executor can have it's move-ability and copy-ability deleted, at least until we find a use case for move-ability of Executors.